[4.7/4.8 Regression] GCC cannot handle array initialization of string constant with point arithmetic properly

 http://gcc.gnu.org/bugzilla/show_bug.cgi?id=53084

--- gcc-4_6-branch/gcc/ChangeLog	2012/04/24 06:10:53	186745
+++ gcc-4_6-branch/gcc/ChangeLog	2012/04/24 06:14:37	186746
@@ -1,6 +1,13 @@
 	benefit speed path.
 	(execute_pre): Use flag_tree_partial_pre.
 
+2012-04-24  Jakub Jelinek  <jakub@redhat.com>
+
+	PR middle-end/53084
+	* varasm.c (compute_reloc_for_constant): Handle ADDR_EXPR
+	of MEM_REF.
+	(output_addressed_constants): Likewise.
+
 2012-04-10  John David Anglin  <dave.anglin@nrc-cnrc.gc.ca>
 
 	PR middle-end/52894
--- gcc-4_6-branch/gcc/testsuite/ChangeLog	2012/04/24 06:10:53	186745
+++ gcc-4_6-branch/gcc/testsuite/ChangeLog	2012/04/24 06:14:37	186746
@@ -1,6 +1,11 @@
 	* gcc.dg/torture/pr51071.c: New testcase.
 	* gcc.dg/torture/pr51071-2.c: Likewise.
 
+2012-04-24  Jakub Jelinek  <jakub@redhat.com>
+
+	PR middle-end/53084
+	* gcc.c-torture/execute/pr53084.c: New test.
+
 2012-04-03  Jason Merrill  <jason@redhat.com>
 
 	PR c++/52796
--- /dev/null
+++ gcc-4_6-branch/gcc/testsuite/gcc.c-torture/execute/pr53084.c	2012/04/24 06:14:37	186746
@@ -0,0 +1,18 @@
+/* PR middle-end/53084 */
+
+extern void abort (void);
+
+__attribute__((noinline, noclone)) void
+bar (const char *p)
+{
+  if (p[0] != 'o' || p[1] != 'o' || p[2])
+    abort ();
+}
+
+int
+main ()
+{
+  static const char *const foo[] = {"foo" + 1};
+  bar (foo[0]);
+  return 0;
+}
--- gcc-4_6-branch/gcc/varasm.c	2012/04/24 06:10:53	186745
+++ gcc-4_6-branch/gcc/varasm.c	2012/04/24 06:14:37	186746
@@ -1,7 +1,7 @@
 /* Output variables, constants and external declarations, for GNU compiler.
    Copyright (C) 1987, 1988, 1989, 1992, 1993, 1994, 1995, 1996, 1997,
    1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
-   2010, 2011  Free Software Foundation, Inc.
+   2010, 2011, 2012  Free Software Foundation, Inc.
 
 This file is part of GCC.
 
@@ -3944,6 +3944,13 @@
 	   tem = TREE_OPERAND (tem, 0))
 	;
 
+      if (TREE_CODE (tem) == MEM_REF
+	  && TREE_CODE (TREE_OPERAND (tem, 0)) == ADDR_EXPR)
+	{
+	  reloc = compute_reloc_for_constant (TREE_OPERAND (tem, 0));
+	  break;
+	}
+
       if (TREE_PUBLIC (tem))
 	reloc |= 2;
       else
@@ -4012,6 +4019,9 @@
 
       if (CONSTANT_CLASS_P (tem) || TREE_CODE (tem) == CONSTRUCTOR)
 	output_constant_def (tem, 0);
+
+      if (TREE_CODE (tem) == MEM_REF)
+	output_addressed_constants (TREE_OPERAND (tem, 0));
       break;
 
     case PLUS_EXPR:
